File: elixir/problem3.exs
#!/usr/bin/env elixir

# Problem 3:
# The prime factors of 13195 are 5, 7, 13 and 29.
# What is the largest prime factor of the number 600851475143 ?

Code.load_file("factor.ex")

defmodule Euler3 do
  def largest_prime_factor(n) do
    n
    |> Factor.factorize()
    |> Enum.reject(&Factor.prime_filter/1)
    |> Enum.sort()
    |> Enum.reverse()
    |> hd
  end
end

IO.inspect(Euler3.largest_prime_factor(600851475143))

File: python/problem3.py
#!/usr/bin/env python2.5

"""
Problem 3 from Project Euler

What is the largest prime factor of the number 600851475143?
"""

from math import sqrt

n = 600851475143

pfact = []

def factors(n):
    f = []
    i = 1

    while True:
        if i > int(round(sqrt(n))): break
        if n % i == 0: f.append(i)
        i += 1

    return f

#print "Possible Prime Factors:", factors(n)

def hasfactors(n):
    if len(factors(n)) == 2:
        return True
    else:
        return False

f = factors(n)
for i in reversed(f):
    if not hasfactors(i):
        print i
        break
##        pfact.append(i)
##
##print pfact[-1]
